tracker.go

```diff
@@ -32,98 +34,160 @@ func (c *Config) SetDefault() {
 	c.thresholdStopRate = 4000
 	c.interval = 5 * time.Second
 	c.reportInterval = 15 * time.Second
-	c.captureMod = 1000
+	c.captureInterval = 200 * time.Millisecond
 	c.writer = os.Stdout
 }
 
 func (c *Config) NewTracker() *StackTracker {
-	return &StackTracker{
+	t := &StackTracker{
 		config: c,
 	}
+	//go t.intervalTester()
+	return t
 }
 
 type StackTracker struct {
-	config             *Config
-	intervalStartTime  time.Time
-	lastReport         time.Time
-	intervalStartCount int32
-	count              int32
-	traces             map[string]uint32
-}
+	config *Config
+	count  int32
 
-func (t *StackTracker) TrackCall(now time.Time) {
-	if t.intervalStartTime.IsZero() {
-		t.intervalStartTime = now
-	}
-	t.count += 1
+	tracesMutex  sync.RWMutex
+	traceChannel chan string
 
-	deltaT := now.Sub(t.intervalStartTime)
-	deltaC := t.count - t.intervalStartCount
+	lastCaptureMutex sync.Mutex
+	lastCapture      time.Time
+}
 
-	if deltaT > t.config.interval {
-		var rate float64 = float64(deltaC) * float64(time.Second) / float64(deltaT)
-		if t.isTracking() {
-			if rate < t.config.thresholdStopRate {
-				t.stopTracking(now)
-			}
-		} else {
-			if rate > t.config.thresholdStartRate {
-				t.startTracking(now)
-			}
+func (t *StackTracker) TrackCall(now time.Time) {
+	for {
+		count := atomic.LoadInt32(&t.count)
+		if atomic.AddInt32(&t.count, 1) == count+1 {
+			break
 		}
-
-		t.intervalStartTime = now
-		t.intervalStartCount = t.count
 	}
 
 	if t.isTracking() {
-		if deltaC%t.config.captureMod == 0 {
-			t.captureTrace()
+		t.lastCaptureMutex.Lock()
+		doCapture := now.Sub(t.lastCapture) > t.config.captureInterval
+		if doCapture {
+			t.lastCapture = now
 		}
-		if now.Sub(t.lastReport) < t.config.reportInterval {
-			t.writeReport(now)
+		t.lastCaptureMutex.Unlock()
+
+		if doCapture {
+			t.captureTrace()
 		}
 	}
 }
 
 func (t *StackTracker) captureTrace() {
-	var pc []uintptr = make ([]uintptr, 64)
+	var pc []uintptr = make([]uintptr, 64)
 	l := runtime.Callers(2, pc)
 	pc = pc[:l]
 	key := fmt.Sprintf("%X", pc)
 	// trim brackets from string of form "[0123 4567 89ab cdef]"
-	key = key[1:len(key)-1]
-	t.traces[key]+= 1
+	key = key[1 : len(key)-1]
+
+	t.tracesMutex.RLock()
+	if t.traceChannel != nil {
+		t.traceChannel <- key
+	}
+	t.tracesMutex.RUnlock()
+
+}
+
+func (t *StackTracker) intervalTester() {
+	intervalStartTime := time.Now()
+	intervalStartCount := int32(0)
+	var finalizeReporter chan interface{}
+	for {
+		time.Sleep(t.config.interval)
+		now := time.Now()
+		deltaT := now.Sub(intervalStartTime)
```

Comment:
That functionality was moved out of TrackCall into a separate goroutine running intervalTester()
Now instead of checking for the interval when we get a trace, we check on a timer.